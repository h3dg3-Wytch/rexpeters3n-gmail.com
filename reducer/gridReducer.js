export const initialState = {
    squares: Array(9).fill(null),
    isCurrentlyX: true
};

export function reducer(state, {type, payload}) {
  switch (type) {
    case 'reset':
      return initialState;
    case 'click':
      if (state.squares[payload.index] != null || payload.winner != null) {
        return state;
      }
      const updatedSquares = state.squares.slice();
      updatedSquares[payload.index] =  getCurrentSymbol(state.isCurrentlyX);
      return { ...state, squares: updatedSquares, isCurrentlyX: !state.isCurrentlyX };
    default:
      return state;
  }
}

export const getWinner = squares => {
    const possibleLines = [
      [0, 1, 2],
      [3, 4, 5],
      [6, 7, 8],
      [0, 3, 6],
      [1, 4, 7],
      [2, 5, 8],
      [0, 4, 8],
      [2, 4, 6]
    ];
    for (let i = 0; i < possibleLines.length; i++) {
      const [a, b, c] = possibleLines[i];
      if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
        return squares[a];
      }
    }
    return null;
  }

  export const isBoardFull = squares => {
    for (let i = 0; i < squares.length; i++) {
      if (squares[i] == null) {
        return false;
      }
    }
    return true;
  }

  export const getStatus = (winner, isBoardFull, isCurrentlyX) => {
    if (winner) {
      return "Winner: " + winner;
    } else if (isBoardFull) {
      return "Draw";
    } else {
      return "Current player: " + (isCurrentlyX ? "X" : "O");
    }
  }

  export const getCurrentSymbol = isCurrentlyX => isCurrentlyX ? 'X' : '0';