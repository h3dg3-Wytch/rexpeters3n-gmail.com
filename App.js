import React, { useState, useReducer } from 'react';
import { StyleSheet, Text, View, Button, TouchableOpacity } from 'react-native';
import Square from './components/Square';
import RejectionButton from './components/RejectionButton';
import Header from './components/Header'

import { reducer, initialState, getWinner, isBoardFull, getStatus } from './reducer/gridReducer';

import { ticTacToeAction, reset } from './actions/gameActions';
import Game from './components/Game';

export default function App() {

  const [state, dispatch] = useReducer(reducer, initialState);

  const resetOnClick = () => reset(dispatch);
  const ticTacToeClick = (index, winner) => ticTacToeAction(dispatch, index, winner);

  const { squares, isCurrentlyX } = state;
  const winner = getWinner(squares);
  const boardFull = isBoardFull(squares);
  const status = getStatus(winner, boardFull, isCurrentlyX);

  return (
    <Game status={status} resetOnClick={resetOnClick} squares={squares} ticTacToeClick={ticTacToeClick} winner={winner} />
  );
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'column', 
    justifyContent: 'center',
    alignItems: 'center',
    height: '100%',
  },
  game: {
    marginTop: 10,
    display: 'flex',
    flexDirection: 'column'
  },
  boardRow: {
    flexDirection: 'row'
  },
  gameStatus: {
    marginTop: 20,
    width: '30%',
    color: 'red'
  },
  restartButton: { 
    height: 20,
    marginTop: 10,
    width: 100,
    justifyContent:'center', 
    alignItems:'center', 
    backgroundColor: 'red' 
  },
});
