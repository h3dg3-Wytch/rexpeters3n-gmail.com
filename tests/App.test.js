import React from 'react';
import renderer from 'react-test-renderer';

import App from '../App';

import { shallow } from 'enzyme';

import { initialState } from '../reducer/gridReducer';

describe('<App />', () => {

  it('can render', () => {
    const component = shallow(<App />);
    expect(component.exists()).toEqual(true);

  });
  
});