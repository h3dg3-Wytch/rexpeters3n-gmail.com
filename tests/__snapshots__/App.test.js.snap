// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> has 2 children 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
        "marginBottom": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 40,
        }
      }
    >
      Current player: X
    </Text>
  </View>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#add8e6",
            "display": "flex",
            "height": 120,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 1,
            "padding": 10,
            "width": 120,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#add8e6",
            "display": "flex",
            "height": 120,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 1,
            "padding": 10,
            "width": 120,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#add8e6",
            "display": "flex",
            "height": 120,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 1,
            "padding": 10,
            "width": 120,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#add8e6",
            "display": "flex",
            "height": 120,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 1,
            "padding": 10,
            "width": 120,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#add8e6",
            "display": "flex",
            "height": 120,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 1,
            "padding": 10,
            "width": 120,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#add8e6",
            "display": "flex",
            "height": 120,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 1,
            "padding": 10,
            "width": 120,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#add8e6",
            "display": "flex",
            "height": 120,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 1,
            "padding": 10,
            "width": 120,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#add8e6",
            "display": "flex",
            "height": 120,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 1,
            "padding": 10,
            "width": 120,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#add8e6",
            "display": "flex",
            "height": 120,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 1,
            "padding": 10,
            "width": 120,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 40,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "red",
            "borderRadius": 24,
            "height": 40,
            "justifyContent": "center",
            "marginTop": 10,
            "opacity": 1,
            "width": 100,
          }
        }
      >
        <Text
          onPress={[Function]}
        >
          Restart
        </Text>
      </View>
    </View>
  </View>
</View>
`;
